Proposal: Multiple user script worlds (#560)
hober/web-extensions: proposals/multiple_user_script_worlds.md

user scripts that execute on the same page; however, these scripts are
conceptually distinct from one another and should generally not interact with
each other.

Today, user script managers take various steps to isolate user scripts as they
are injected, since all user scripts will run either in the context of the
main world or in a single user script world for the extension. This API would
allow user script managers to instead have multiple user script worlds, each
isolated from one another, to reduce the likelihood of user scripts negatively
affecting one another through collisions.

#### Use Cases

The primary use case is user script managers that handle multiple, independent
script injections in a single page.

### Known Consumers

We anticipate user script managers such as Tampermonkey, Violentmonkey, and
others to use this new capability.

## Specification

### Schema

This change would modify existing schemas. We will add a new property,
`worldId`, to the `WorldProperties` and `RegisteredUserScript` types, as below.

Relevant methods and types:

```diff
 export namespace userScripts {
   export interface WorldProperties {

>      /**

Reviewer: ```suggestion
+    /**
```

Author: Done

      * injected into matching pages.
      */
     js: ScriptSource[];

     /**
      * Specifies which pages this user script will be injected into. See
      * Match Patterns for more details on the syntax of these strings. This
      * property must be specified for ${ref:register}.
      */
     matches?: string[];

     /**
      * Specifies when JavaScript files are injected into the web page. The
      * preferred and default value is document_idle
      */
     runAt?: RunAt;

     /**
      * The JavaScript execution environment to run the script in. The default
      * is `USER_SCRIPT`
      */
     world?: ExecutionWorld;

+    /**
+     * If specified, specifies a specific user script world ID to execute in.
+     * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
+     * script will execute in the default user script world.
+     * Values with leading underscores (`_`) are reserved.
+     */
+    worldId?: string;
   }
 }
```

Note that the signatures of the related functions, including `configureWorld()`,
`register()`, and others are left unchanged.

When the developer specifies a `worldId` in either the `WorldProperties` or

> a `RegisteredUserScript`, the browser will create a separate user script world

Reviewer: Important design question: what are the persistence/lifetime characteristics of worldId + configureWorld? worldId should not persist across extension updates for consistency with the existing userScripts API design (https://developer.chrome.com/docs/extensions/reference/api/userScripts#extension_updates), but there is still an open question of the lifetime of worldId (and consequently, potentially reduce the duration of persistence).

Previously, `configureWorld` would have a limited scope, exactly one. Now, with the introduction of `worldId`, there may be an unbounded number. According to the current proposal, the concept of a specific world (identified by worldId) becomes concrete whenever it is used in `RegisteredUserScript` or `WorldProperties`. Because the number of worlds is now dynamic, we should clarify when a worldId exists and when it ceases to exist. As currently written, worldIds are instantiated and never discarded.

This section seems to have been written with `userScripts.register` in mind. The issue that I'm thinking of would have become much more obvious if the proposal explicitly spelled out the expected behavioral change (or non-change) per method (from the proposal process perspective - I think that it would be a good idea for future proposals to exhaustively enumerate all uses of a type to make sure that the behavior is fully specified where needed):
- `userScripts.register()` - may introduce a worldId. Does not necessarily have to do anything until the script is used in the renderer (unless there is a desire to broadcast all known worldIds?).
- `userScripts.configureWorld()` - may create a persistent configuration for a world. What if the worldId does not exist yet?
- `userScripts.getScripts()` - does this have to "create a separate user script world"?
- `userScripts.unregister()` - when the last use of `worldId` disappears, would that delete the configuration?
- `userScripts.update()` - when the `worldId` changes, and the previous one was the last one, does that delete the registered configured worlds?

I think that there are roughly two ways to address this issue (not mutually exclusive):

1. Explicit world management (`userScripts.registerWorld` + `userScripts.unregisterWorld`). Using an unregistered `worldId` in RegisteredUserScript or WorldProperties is an error.
2. Automatic world cleanup: a world configuration is guaranteed to persist when there is at least one RegisteredUserScript that references it. What happens when there is no reference to it is more interesting:
   - When there are no references, we could unregister the world as soon as the refcount drops to zero would be obvious (by `userScripts.unregister` or `userScripts.update`).
   - Do we want to support calling `userScripts.configureWorld` before `userScripts.register`? If so, then `configureWorld` should register a world implicitly, despite there not being any scripts that reference it.
     - to avoid accumulating junk, unused worlds should eventually be discarded. This is non-trivial.
     - an alternative is to enable the world to be configured along with registering a user script.
     - another alternative is to not support worldId in `configureWorld` unless the worldId has been allocated explicitly by `userScripts.register`. Extensions that really care about ensuring setup of the world before execution can register a dummy script with a worldId, then configure the world, then register the actual script for that world, and then unregister the dummy script. Is this an acceptable pattern?

Author: Thanks, @Rob--W  !

I was thinking about most of these already (complete with a few TODOs, e.g. [here](https://chromium-review.googlesource.com/c/chromium/src/+/5408252/4/extensions/browser/api/user_scripts/user_scripts_api.cc#565)) and was planning on submitting an update to the proposal as we got nearer, but I hadn't considered the update behavior and allowing developers to clear out old registrations -- both of which I agree are valuable.

The way I'm thinking about worlds is that:
- A user script is registered with a world.  That determines where it injects.
- The configuration for that world may or may not exist, depending on whether the developer wants to use non-default values (for messaging and CSP, but potentially other APIs in the future).

As such, there isn't really the "creation" of a world when you register a script -- worlds are created in the document in which they run, and are created according to the configuration provided (if any).

I don't like the idea of either reference counting or requiring a user script be registered for every specified worldId -- that sounds very difficult for both us (the browser) and for developers to keep track of, and seems like it would very likely lead to some unpredictable or unexpected edge cases (e.g., a developer updates a user script world via unregister + register, and doesn't realize that unregistered the user script world).  It also leads to a (very small) race where it'd be impossible to ensure a user script injected with given world properties, since you'd have to register the script _before_ the configuring the world, but then the script may inject before that configuration took hold.  I'd much rather lean on the developers handling the registration.

I've expanded the proposal to include:
- Two new functions, `removeWorld()` and `getWorldConfigurations()`.
- Notes about how the worlds behave when injecting.
- Notes on world persistence and world limits.

Please take a look and let me know what you think.

      * injected into.
      */
     includeGlobs?: string[];

     /**
      * The list of ScriptSource objects defining sources of scripts to be
      * injected into matching pages.
      */
     js: ScriptSource[];

     /**
      * Specifies which pages this user script will be injected into. See
      * Match Patterns for more details on the syntax of these strings. This
      * property must be specified for ${ref:register}.
      */
     matches?: string[];

     /**
      * Specifies when JavaScript files are injected into the web page. The
      * preferred and default value is document_idle
      */
     runAt?: RunAt;

     /**
      * The JavaScript execution environment to run the script in. The default
      * is `USER_SCRIPT`
      */
     world?: ExecutionWorld;

+    /**
+     * If specified, specifies a specific user script world ID to execute in.
+     * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
+     * script will execute in the default user script world.
+     * Values with leading underscores (`_`) are reserved.
+     */
+    worldId?: string;
   }

   ...

+   /**
+    * Removes the configuration for a given world.
+    */

> +   export function removeWorld(worldId: string): Promise<void>;

Reviewer: "Remove world" gives the impression that the world ceases to exist. The way I see it, the concept of a world (identified by worldId) starts to exist whenever it is used by a script or config. A world is only removed when all scripts referencing the worldId are unregistered (AND the config removed too).

So I suggest renaming to something more obvious, such as one of the following:

- removeWorldConfiguration
- resetWorldConfiguration
- clearWorldConfiguration

And can you also clarify the behavior when the worldId doesn't have a special configuration? Especially with the reset or clear name, I would not expect errors. I would be fine with any defined behavior.

Author: That's a good point on the naming; thanks, Rob.  "clearWorldConfiguration" SGTM, so I've tentatively incorporated that (if anyone has strong preferences for "reset", I'm good with that one, too).

+
+   /**
+    * Returns a promise that resolves to an array of the the configurations
+    * for all user script worlds.
+    */
+   export function getWorldConfigurations(): Promise<WorldProperties[]>;
 }
```

Note that the signatures of the related functions, including `configureWorld()`,
`register()`, and others are left unchanged.

When the developer specifies a `RegisteredUserScript`, the browser will use a
separate user script world when injecting the scripts into a document. If
`worldId` is omitted, the default user script world will be used.

Worlds may be configured via `userScripts.configureWorld()` by indicating the
given `worldId`. User scripts injected into a world with the given `worldId`
will have the associated properties from the world configuration. If a world
does not have a corresponding configuration, it uses the default user script

> world properties.

Reviewer: Should also add that configuration changes only apply to worlds/execution contexts created after that point. Existing worlds stick to the previous configuration.

If you envision different behavior, please speak up.

Author: This one is a bit tricky.

I don't think we'd:
- Update the CSP of existing worlds
- Manipulate JS objects in the existing worlds

However, I do think we _may_ prevent APIs from working if they get proxied through the browser process.  In particular, I think we may want to prevent messaging APIs from working (the browser may see that the user script world does not currently allow messaging, and prevent any message communication).  This seems desirable to me, and is also inline with the behavior we have for e.g. when the extension is unloaded but content scripts continue to run: the JS is unaffected, but anything reaching _outside_ that world (e.g. through the browser process) may be.

I've added this in.

World configurations can be removed via the new `userScripts.removeWorld()`
method.

Additionally, `runtime.Port` and `runtime.MessageSender` will each be extended
with a new, optional `userScriptWorldId` property that will be populated in the
arguments passed to `runtime.onUserScriptConnect` and
`runtime.onUserScriptMessage` if the message initiator is a non-default user
script world. We do not need to populate a value for messages from the default
user script world, since this is unambiguous given the distinction in the
event listeners (`onUserScriptMessage` already indicates the message was from
a user script).

### Behavioral Notes

#### World Persistence

Configured worlds are persisted until the owning extension is updated (in order
to align with the behavior of the rest of the `userScripts` API) or manually
removes them via the new `userScripts.removeWorld()` method.

#### World Limits

> User agents may place limits on both the number of registered worlds and the

Reviewer: Should we really limit the number of worlds? What are the numbers you are thinking of?

A too low limit reduces the practical value of the multiple worlds concept, and user script managers are then still forced to implement script wrappers to avoid conflicts between user scripts.

A reasonably high limit reduces the impact for users with a few scripts, but punishes power users with lots of scripts. If a user is willing to install many scripts and considers the overhead acceptable, why would we stop them?

Reviewer 2: Assuming the world's overhead is ~1MB, how about 256 worlds max?

Author: Short answer: yes, I strongly think we should limit the number of worlds.  In particular, I think this is true for the active number of worlds on a page.

The registration is relatively cheap, so I don't think we need to be _too_ limiting for total number of registered worlds (I was originally thinking ~100, but I think that one's more flexible).  The real cost is in the number of active worlds on pages.  Each of these results in a separate JS context which, depending on the size of the script, may (very) non-trivially increase the weight of a given document.  Then, given the number of documents that can be actively running, this can become _much_ heavier weight.  Every additional user script world is roughly equivalent to the cost of a whole other content script-injecting extension.  From @tophf 's suggestion of 256 worlds assuming a 1MB overhead, that's 256MB _per document_.  (Assuming that was meant as a suggestion for "active worlds on a page", rather than total registered worlds.  I think 256 total registered worlds is perhaps a bit high, but not unreasonable.)  I just opened cnn.com (my go-to for a "noisy" site), and there are 30 different frames there.  That means that, with 256 worlds at 1MB a piece, the user would have 7.5GB of memory used on _that one tab_ just from user scripts.  I think that's more than we'd allow in Chrome.

Now, since a lot of user scripts are only intended to run on specific pages, there are lots of cases when two scripts would never "collide" and result in multiple worlds on the same page, so I'm hopeful that most user script managers wouldn't run into this limit frequently.  But for all the scripts that inject on every page, I think there is a risk of major performance impact.

I also think the exact values here may be something left up to individual browser vendors to establish, as we've done with historical limitations with performance (in which case, Firefox may have no limit, Safari on iOS may establish a different limit than Safari on MacOS, etc).  But rather than fully leave out the expectations from the specification and then have an individual browser impose an intervention in ways that are unpredictable to the developer, I think it's worth describing what the behavior the developer can expect will be.

Reviewer 2: Chrome/Firefox similarly limit the number of OS processes they use and they combine sites in one process if necessary. Arguably extensions could set the maximum number of the worlds or at least read it from the API so they can display the warning to the user and add a "virtualizer" to the userscript code, specifically `with (globalProxy) {....}`, which is undesirable as `with` incurs overhead.

FWIW, my 1MB estimate might be too high. What if it's just 100kB?

Reviewer 2: I see 46kB for an extension `Window` in Chrome Devtools -> Memory -> Snapshot.

Roughly 20 bytes \* (1000 global objects + their 1000 prototypes).

Author: Unfortunately, it's not _just_ the overhead of the JS items -- there's more that needs to happen to create a new isolated world (on the Chrome side, we need a new v8::Context, ScriptState, execution environment, etc).  The impact I'm seeing for a completely empty script injection is about 200KB per world when injecting into a very simple site (example.com).  Because the isolated world also has a copy of the DOM (the DOM is shared, but the JS objects for the DOM are not), this increases with the complexity of the page.

I am supportive of exposing the maximum number of worlds so that extensions can handle this appropriately (as @tophf called out, potentially showing user-visible UI or modifying their injections to be "semi-isolated" from one another).

Reviewer 2: So what happens to a matched userscript in the case its separate world cannot be created? There should be some event or another method of detecting it.

Author: This is called out in the proposal -- it will fallback to using the default world.

I'd be happy to consider:
- Having an additional trait in WorldProperties like "shouldFallBackIfWorldLimitReached" (perhaps with a better name) that controls whether the script should fall back or just bail and not inject, and / or
- Having a new event like "onUserScriptCouldNotInjectBecauseWorldLimitReached" (definitely with a better name)

However, given the scope creep we're seeing in this proposal, I'd like to do those in a separate change.

Reviewer 2: ...because the userscript manager extension would need to run the userscript's function inside `with(windowProxy)`

Reviewer 2: Okay, as long as it's not postponed to some distant future.

Author: I think this can be a reasonably fast-follow.

Reviewer: I'm not resolving the thread yet because the expressed questions has not been resolved, but feel free to merge even if this thread is still open.

Consider calling out improvements in this area in "future work".

Reviewer 2: An event is necessary anyway, because managers need it to set up the script's environment before execution of any matched userscript e.g. set `window[randomId]` with the secure communication channel name in the MAIN world right right before execution so that no JS code can run in-between. Something like chrome.userScripts.onExec with (scriptId, worldId) parameters and chrome.runtime.lastError inside the listener.

Reviewer: Synchronous cross-world communication and code execution has been discussed before, including recently at the in-person WECG meetup. That will become a separate proposal, and does not need to be specific to user scripts.

Reviewer 2: My point was that userscripts need an event right before execution, it's not the same as cross-world communication. This event can be used to report the world creation error.

Reviewer 2: Same-world userscripts in the default world need it too to set up the internal bindings for `GM` stuff which is implemented in a common controller script. BTW we need chrome.userScripts.registerController for the code that runs before the first matched userscript in the tab and receives all matched userscript ids.

Author: Sounds like we're agreed we can discuss these events in a follow-up proposal, so I'll resolve this conversation for now.

number of worlds that may be active in a given document. These limits may be
different (since an extension may want more individual world configurations than
they would expect to be practically encountered on a given site).

> If an extension reaches the limit of the number of registered worlds, attempts

Reviewer: "registered worlds" -> "registered world configurations"

World registration is ambiguous because there are two distinct ways to have a concept of a world. And I think that you are trying to encourage extensions to unregister obsolete world configurations here.

Author: Done

# Proposal: Allow multiple user script worlds

**Summary**

Allow developers to configure and use multiple user script worlds in the
`browser.userScripts` API.

**Document Metadata**

**Author:** rdcronin

**Sponsoring Browser:** Google Chrome

**Contributors:** N/A

**Created:** 2024-03-07

> **Related Issues:** TODO

Author: @Rob--W I didn't see an existing bug for this.  Do you know if one is filed?

Reviewer: Multiple world has been mentioned from the start in the userScripts API discussions at https://github.com/w3c/webextensions/issues/279

E.g. fragment from https://github.com/w3c/webextensions/issues/279#issuecomment-1255843961

> At present, I propose only one new world. This is strictly an improvement over the current state of art. Support for more than one world could be useful to extensions, but raises the complexity of implementation, so I'm leaving that for a follow-up and consider it to be out of scope in this proposal.

Author: I think this would be better tracked individually (rather than lumped into the meta "user scripts in MV3" issue), so I've filed a new issue ( issue #565 ) for it, but referenced the original comment.

## Motivation

### Objective

User scripts are (usually small) scripts that enable myriad use cases by
injecting scripts into different web pages. A user may have many different
user scripts that execute on the same page; however, these scripts are
conceptually distinct from one another and should generally not interact with
each other.

Today, user script managers take various steps to isolate user scripts as they
are injected, since all user scripts will run either in the context of the
main world or in a single user script world for the extension. This API would
allow user script managers to instead have multiple user script worlds, each
isolated from one another, to reduce the likelihood of user scripts negatively
affecting one another through collisions.

#### Use Cases

The primary use case is user script managers that handle multiple, independent
script injections in a single page.

### Known Consumers

We anticipate user script managers such as Tampermonkey, Violentmonkey, and
others to use this new capability.

## Specification

### Schema

This change would modify existing schemas.

Relevant methods and types:

```

> export namespace userScripts {

Reviewer: Instead of copy-pasting the schema and letting the reader eyeball-diff it, could you just specify the final schema, and mention what changed compared to before? And consider linking the original proposal in this repo.

Author: I prefer having the original schema so that, when looking back at these in <n> years, we can easily see what changed.

I already explicitly called out the changed lines with "+" indicators, but I've called out explicitly that I've done that.

Reviewer: Markdown supports diff rendering, by putting "diff" after the triple backtick. Then when every line is indented with one space, a + instead of a space at the start renders as an addition and - as a removal.

```diff
 Line unchanged
+Line added
-Line removed
```

Author: Ooh, that's rad!  Thanks for the tip; done.

With that, I think it's also clear to condense these into one section, so did that as well.

  export interface WorldProperties {
    /**
     * Specifies the world csp. The default is the `` `ISOLATED` `` world csp.
     */

>     csp?: string;

Reviewer 3: Ditto regarding #563.

     */
    messaging?: boolean;
  }

  /**
   * Properties for a registered user script.
   */
  export interface RegisteredUserScript {
    /**
     * If true, it will inject into all frames, even if the frame is not the
     * top-most frame in the tab. Each frame is checked independently for URL
     * requirements; it will not inject into child frames if the URL
     * requirements are not met. Defaults to false, meaning that only the top
     * frame is matched.
     */
    allFrames?: boolean;
    /**
     * Specifies wildcard patterns for pages this user script will NOT be
     * injected into.
     */
    excludeGlobs?: string[];
    /**
     * Excludes pages that this user script would otherwise be injected into.
     * See Match Patterns for more details on the syntax of these strings.
     */
    excludeMatches?: string[];
    /**
     * The ID of the user script specified in the API call. This property
     * must not start with a '_' as it's reserved as a prefix for generated
     * script IDs.
     */
    id: string;
    /**
     * Specifies wildcard patterns for pages this user script will be

>      * injected into.

Reviewer: Although the behavior of includeGlobs/matches did not change, I'd like to emphasize that the original proposal explicitly called out the expected matching behavior at https://github.com/w3c/webextensions/blob/14ebd85ac034e5e4138d0653ae797a4aa9dc6703/proposals/user-scripts-api.md#L66-L67 upon my request, because the expected glob/matches behavior from user scripts differs from what has been implemented in the content script APIs.

Why is the original note omitted from the current "full" API description?

Author: Good question, but orthogonal from this issue.  I can update that separately in Chrome's repo.

     */
    includeGlobs?: string[];
    /**
     * The list of ScriptSource objects defining sources of scripts to be
     * injected into matching pages.
     */
    js: ScriptSource[];
    /**
     * Specifies which pages this user script will be injected into. See
     * Match Patterns for more details on the syntax of these strings. This
     * property must be specified for ${ref:register}.
     */
    matches?: string[];
    /**
     * Specifies when JavaScript files are injected into the web page. The
     * preferred and default value is document_idle
     */
    runAt?: RunAt;
    /**
     * The JavaScript execution environment to run the script in. The default
     * is `USER_SCRIPT`
     */

>     world?: ExecutionWorld;

Reviewer 3: I filed #563 about this, but IMO we should use lowercase strings. This has recently changed with these newer APIs, while older APIs use lowercase strings — leading to an inconsistency.

Author: Thanks for raising this.  I agree with aligning on a uniform standard, but would prefer using all caps for the reasons I highlighted in the response there.

I'd like to avoid blocking this issue on the resolution of that (it seems like it is very likely to have lots of impassioned bikeshed discussions : )); for the purpose of this proposal, I'd propose we keep the all caps since that aligns with the rest of the user script API, and we can establish what to do for other cases as part of issue #563  (since we'll need to address other APIs anyway).  Does that sound reasonable to you?

  }
}
```

We will add a new property, `worldId`, to the `WorldProperties` and
`RegisteredUserScript` types, as below.

```
export namespace userScripts {
  export interface WorldProperties {
    /**
+    * Specifies the ID of the specific user script world to update.
+    * If not provided, updates the properties of the default user script world.
+    */
+   worldId?: string;

    /**

>      * Specifies the world csp. The default is the `ISOLATED` world csp.

Reviewer: ```suggestion
     * Specifies the world's CSP. The default is the `ISOLATED` world's CSP.
```

Author: Done

     */
    csp?: string;

    /**
     * Specifies whether messaging APIs are exposed. The default is `false`.
     */
    messaging?: boolean;
  }

  /**
   * Properties for a registered user script.
   */
  export interface RegisteredUserScript {
    /**
     * If true, it will inject into all frames, even if the frame is not the
     * top-most frame in the tab. Each frame is checked independently for URL
     * requirements; it will not inject into child frames if the URL
     * requirements are not met. Defaults to false, meaning that only the top
     * frame is matched.
     */
    allFrames?: boolean;

>     /**

Reviewer 3: Consider adding empty lines before all these comments, you do in some places and it helps readability.

Author: Done

     * See Match Patterns for more details on the syntax of these strings.
     */
    excludeMatches?: string[];
    /**
     * The ID of the user script specified in the API call. This property
     * must not start with a '_' as it's reserved as a prefix for generated
     * script IDs.
     */
    id: string;
    /**
     * Specifies wildcard patterns for pages this user script will be
     * injected into.
     */
    includeGlobs?: string[];
    /**
     * The list of ScriptSource objects defining sources of scripts to be
     * injected into matching pages.
     */
    js: ScriptSource[];
    /**
     * Specifies which pages this user script will be injected into. See
     * Match Patterns for more details on the syntax of these strings. This
     * property must be specified for ${ref:register}.
     */
    matches?: string[];
    /**
     * Specifies when JavaScript files are injected into the web page. The
     * preferred and default value is document_idle
     */
    runAt?: RunAt;
    /**
     * The JavaScript execution environment to run the script in. The default
     * is `USER_SCRIPT`
     */

>     world?: ExecutionWorld;

Reviewer 3: Ditto regarding #563.

+   /**
+    * If specified, specifies a specific user script world ID to execute in.
+    * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
+    * script will execute in the default user script world.
+    */

> +   worldId?: string;

Reviewer: Are there any constraints on the format of `worldId`? For example, if we reserve `_` for internal use, then we can choose default worldIds as follows.

- `world: "MAIN"` -> `worldId` defaults to `_MAIN_WORLD` and an error raised if anything else is passed.
-  `world: "USER_SCRIPT"` -> `worldId` defaults to `_USER_SCRIPT_WORLD` and can be set to anything else.

Another option is to use `""` as the default value, and permit any value in `worldId`.

Author: I was debating that.  I don't want to surface "_" to developers (so wouldn't want them to have to pass  "_DEFAULT" or anything), but I'm fine saying _-prefixed values are reserved, allowing us to use these internally (if we want).

  }
}
```

Note that the signatures of the related functions, including `configureWorld()`,
`register()`, and others are left unchanged.

> When the developer specifies a `worldId` in either the `WorldProperties` or

Reviewer: `WorldProperties` was not defined before. For clarity, can you list all methods that are affected by this change.

Author: World properties were (and are) defined.  They allow the configuration of whether messaging APIs are exposed to the world.  In chrome, these are [here](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/api/user_scripts.idl;l=77-87;drc=0345221adefeb4f5f77cb57dd2dd1e8159b30762).

a `RegisteredUserScript`, the browser will create a separate user script world

> for those cases. If `worldId` is omitted, a "default" user script world will

Reviewer: ```suggestion
for those cases. If `worldId` is omitted, the default user script world will
```

Author: Done

worlds, which would have a heavy performance cost. However, this is not a
security consideration. Additionally, browsers can mitigate this by enforcing
a limit of the maximum number of user script worlds an extension may register
or have active at a given time.

These enforcements will be left up to the browser to implement, if they feel
necessary.

### Additional Security Considerations

No additional security considerations. This API may (mildly) increase security
by (somewhat) isolating individual user scripts from interacting with one
another.

## Alternatives

### Existing Workarounds

Today, user script managers go to various lengths to try to isolate behavior
between user scripts. This includes having various frozen types, making heavy
use of function closures, ensuring injection before untrusted code, and more.

These workarounds are fragile and require significantly more complex code in
user script managers, while providing a lesser guarantee of isolation than a
separate user script world would provide.

### Open Web API

The concept of user scripts should not be exposed to the web; this should not
be an open web API.

## Implementation Notes

N/A

> ## Future Work

Reviewer: Would we ever expose `worldId` to the content side? A use case that I can see is enabling (synchronous) cross-world communication or execution. I've wanted to create such a proposal for a while.

Author: I'm not quite sure what you mean by "content side" -- do you mean to the user script?

In either case, I've indicated we may leverage the world ID for inter-world communication.

Reviewer: By "content" I meant in the renderer. Because doing so opens two things: firstly the ability to do interworld communication and two, enabling extensions to use it as a hack to encode a world-specific configuration in the string.

Since you included the mention of the first, the answer appears to be yes and I'll resolve this thread.

With the addition of a `userScripts.execute()` function (as described in
this [PR](https://github.com/w3c/webextensions/pull/540) and

> [issue](https://github.com/w3c/webextensions/issues/477), we should also

Reviewer: ```suggestion
[PR 540](https://github.com/w3c/webextensions/pull/540) and
[issue 477](https://github.com/w3c/webextensions/issues/477), we should also
```

Added number so that readers don't have to hover over the link to see what it is referring to.

Author: Good call; added.

# Proposal: Allow multiple user script worlds

**Summary**

Allow developers to configure and use multiple user script worlds in the
`browser.userScripts` API.

**Document Metadata**

**Author:** rdcronin

**Sponsoring Browser:** Google Chrome

**Contributors:** N/A

**Created:** 2024-03-07

**Related Issues:** [565](https://github.com/w3c/webextensions/issues/565)

## Motivation

### Objective

User scripts are (usually small) scripts that enable myriad use cases by
injecting scripts into different web pages. A user may have many different
user scripts that execute on the same page; however, these scripts are
conceptually distinct from one another and should generally not interact with
each other.

Today, user script managers take various steps to isolate user scripts as they
are injected, since all user scripts will run either in the context of the
main world or in a single user script world for the extension. This API would
allow user script managers to instead have multiple user script worlds, each

> isolated from one another, to reduce the likelihood of user scripts negatively

Reviewer 4: nit: Is it worth clarifying somewhere (in the paragraph here?  in the "Additional Security Considerations" section below?  in a separate doc?) that "isolated" is mostly in "hygiene" sense and not necessarily in "security" sense, because:

* Cross-world contact is possible (e.g. via `window`)
* OS process offers a more robust security boundary (e.g. Spectre allows cross-world reading of the whole process memory)

Reviewer 2: Side-channel attacks notwithstanding, it's much more than hygiene, because isolation means JavaScript artifacts are isolated such as variables and DOM expandos and there's no way to access them directly, same as extension's content scripts are isolated from the MAIN world scripts. The "cross-world contact" is only possible via events and modifying DOM elements, which doesn't violate JavaScript isolation concept.

Reviewer 2: What about `JavaScript environment is isolated from one another (note that a side-channel attack is still possible as it's the same OS process)`?

Author: I think @anforowicz is looking at this from the browser security model, where we don't consider this to be a strict security boundary.  I already had language that hinted at this in the "Additional Security Considerations" (with its "mildly" and "somewhat" language : )), but I've made this more clear by adding an additional note that says this is not a hard security boundary and we wouldn't consider it in the browser's security model.

affecting one another through collisions.

#### Use Cases

The primary use case is user script managers that handle multiple, independent
script injections in a single page.

### Known Consumers

We anticipate user script managers such as Tampermonkey, Violentmonkey, and
others to use this new capability.

## Specification

### Schema

This change would modify existing schemas. We will add a new property,

> `worldId`, to the `WorldProperties` and `RegisteredUserScript` types, as below.

Reviewer 4: Q1: I assume that world ids are generated by the browser (when an extension asks to create a new world or world id)?  (as opposed to having an extension generate unique world id strings, and then pass those IDs to the browser/extension APIs)

Q2: Can a malicious extension try to spoof/forge world ids that belong to another extension?

Author: Q1: not quite -- these IDs _are_ provided by extensions.  On the chrome side, we'll segment these by extension ID, so that the unique world configuration ID would be pair<extension id, world id> (though note that there may be multiple worlds with that same pair across different renderers -- this is intentional, those should inject in the same world configuration).

Q2: Nope, because of this pairing.  (Extension A's "my world" has no overlap with Extension B's "my world").

+    */
+   export function resetWorldConfiguration(worldId: string): Promise<void>;
+
+   /**
+    * Returns a promise that resolves to an array of the the configurations
+    * for all user script worlds.
+    */
+   export function getWorldConfigurations(): Promise<WorldProperties[]>;
 }
```

Note that the signatures of the related functions, including `configureWorld()`,
`register()`, and others are left unchanged.

When the developer specifies a `RegisteredUserScript`, the browser will use a
separate user script world when injecting the scripts into a document. If
`worldId` is omitted, the default user script world will be used.

Worlds may be configured via `userScripts.configureWorld()` by indicating the
given `worldId`. User scripts injected into a world with the given `worldId`
will have the associated properties from the world configuration. If a world
does not have a corresponding configuration, it uses the default user script
world properties. Any existing worlds are not directly affected by
`userScripts.configureWorld()` calls; however, the browser may revoke
certain privileges (for instance, message calls from existing user script worlds

> may beging to fail if the extension sets `messaging` to false). This is inline

Reviewer: ```suggestion
may beging to fail if the extension sets `messaging` to false). This is in line
```

Author: Done

world.

### New Permissions

No new permissions are necessary. This is inline with the `userScripts` API's
current functionality and purpose.

### Manifest File Changes

No manifest file changes are necessary.

## Security and Privacy

### Exposed Sensitive Data

This does not expose any additional sensitive data.

### Abuse Mitigations

This API does not provide any additional avenue for abuse beyond what the
`userScripts` API is already capable of.

Extension developers may abuse this API by requiring too many user script
worlds, which would have a heavy performance cost. However, this is not a
security consideration. Additionally, browsers can mitigate this by enforcing
a limit of the maximum number of user script worlds an extension may register
or have active at a given time.

These enforcements will be left up to the browser to implement, if they feel
necessary.

### Additional Security Considerations

No additional security considerations. This API may (mildly) increase security
by (somewhat) isolating individual user scripts from interacting with one
another.

## Alternatives

### Existing Workarounds

> Today, user script managers go to various lengths to try to isolate behavior

Reviewer 5: Adding references of these examples (stackoverflow link?) would "complete" the document. Assuming the proposal guideline is not against that. This is not a blocking comment.

Author: I've added a link to issue 279, where many of these were discussed.

      * script IDs.
      */
     id: string;

     /**
      * Specifies wildcard patterns for pages this user script will be
      * injected into.
      */
     includeGlobs?: string[];

     /**
      * The list of ScriptSource objects defining sources of scripts to be
      * injected into matching pages.
      */
     js: ScriptSource[];

     /**
      * Specifies which pages this user script will be injected into. See
      * Match Patterns for more details on the syntax of these strings. This
      * property must be specified for ${ref:register}.
      */
     matches?: string[];

     /**
      * Specifies when JavaScript files are injected into the web page. The
      * preferred and default value is document_idle
      */
     runAt?: RunAt;

     /**
      * The JavaScript execution environment to run the script in. The default
      * is `USER_SCRIPT`
      */
     world?: ExecutionWorld;

+    /**
+     * If specified, specifies a specific user script world ID to execute in.
+     * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
+     * script will execute in the default user script world.

> +     * Values with leading underscores (`_`) are reserved.

Reviewer 4: optional nit: Maybe require **two** leading underscores?  It seems that double-leading-underscore is less likely to be used in practice than a single leading underscore.  And this would also be [consistent with C++](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL51-CPP.+Do+not+declare+or+define+a+reserved+identifier#:~:text=Each%20name%20that%20contains%20a,name%20in%20the%20global%20namespace.).

QUESTION: Just to double-check my understanding - reserved world ids may lead to some undesirable behavior **within** an extension, but this is okay, because they won't have any effect **outside** of an extension.  In other words, a malicious extension won't be able to abuse reserved world ids to somehow cross a security boundary somewhere, right?

Author: Thanks, Lukasz!

> optional nit: Maybe require two leading underscores?

Unfortunately, we have a decent amount of precedent in extension APIs for using single-underscore for reserved (a few different APIs, file paths in the extension directory, etc).  We might revisit that at some point in the future, but I think being consistent is best here for now.

> In other words, a malicious extension won't be able to abuse reserved world ids to somehow cross a security boundary somewhere, right?

They certainly shouldn't be able to! ; )  I'd expect any implementation to simply reject any call from an extension that specifies a world ID beginning with '_'.  The reservation of '_' IDs was just a suggestion in case implementors wanted to allow all user script worlds to have a world ID (e.g., using "_default" for the default world in the underlying implementation).  On the Chromium side, we just use a std::optional<>, so there's really no way an extension should be able to abuse this.

This API does not provide any additional avenue for abuse beyond what the
`userScripts` API is already capable of.

Extension developers may abuse this API by requiring too many user script
worlds, which would have a heavy performance cost. However, this is not a
security consideration. Additionally, browsers can mitigate this by enforcing
a limit of the maximum number of user script worlds an extension may register
or have active at a given time.

These enforcements will be left up to the browser to implement, if they feel
necessary.

### Additional Security Considerations

No additional security considerations. This API may (mildly) increase security
by (somewhat) isolating individual user scripts from interacting with one
another.

Note that since isolated world boundaries are not considered "hard" boundaries,
we would not consider this isolation a security measure from the browser's
security model (just as we don't consider isolated worlds to be a strict
security boundary today).

## Alternatives

### Existing Workarounds

Today, user script managers go to various lengths to try to isolate behavior
between user scripts (for more information, see conversation in

> [issue 279](https://github.com/w3c/webextensions/issues/279). This includes

Reviewer: ```suggestion
[issue 279](https://github.com/w3c/webextensions/issues/279)). This includes
```

Added missing closing parenthesis.

Reviewer: Actually, I cannot use Github's UI to commit the change because I don't have write access to the repository from which this PR is pulling. @rdcronin could you apply this change and then merge the PR?

Author: Done!
